devito/builtins.py

```diff
@@ -53,6 +54,131 @@ def smooth(f, g, axis=None):
         dv.Operator(dv.Eq(f, g.avg(dims=axis)), name='smoother')()
 
 
+def gaussian_smooth(f, sigma=1, _order=4, mode='reflect'):
+    """
+    Gaussian smooth function.
+    """
+    class DataDomain(dv.SubDomain):
+
+        name = 'datadomain'
+
+        def __init__(self, lw):
+            super(DataDomain, self).__init__()
+            self.lw = lw
+
+        def define(self, dimensions):
+            return {d: ('middle', self.lw, self.lw) for d in dimensions}
+
+    def fill_data(g, f, dim, lw, mode):
+
+        sl = slice(lw, -lw, 1)
+        indices = []
+        for _ in g.grid.dimensions:
+            indices.append(sl)
+        if isinstance(f, np.ndarray):
+            g.data[tuple(indices)] = f[:]
+        elif isinstance(f, dv.Function):
+            g.data[tuple(indices)] = f.data[:]
+        else:
+            raise NotImplementedError
+        indfl = []
+        indgl = []
+        indfr = []
+        indgr = []
+        if mode == 'constant':
+            for d in g.grid.dimensions:
+                if d == dim:
+                    indfl.append(0)
+                    indgl.append(slice(0, lw, 1))
+                    indfr.append(-1)
+                    indgr.append(slice(-lw, -1, 1))
+                else:
+                    indfl.append(slice(0, -1, 1))
+                    indgl.append(slice(lw, -lw, 1))
+                    indfr.append(slice(0, -1, 1))
+                    indgr.append(slice(lw, -lw, 1))
+            if isinstance(f, np.ndarray):
+                g.data[tuple(indgl)] = f[tuple(indfl)]
+                g.data[tuple(indgr)] = f[tuple(indfr)]
+            elif isinstance(f, dv.Function):
+                g.data[tuple(indgl)] = f.data[tuple(indfl)]
+                g.data[tuple(indgr)] = f.data[tuple(indfr)]
+            else:
+                raise NotImplementedError
+        elif mode == 'reflect':
+            for d in g.grid.dimensions:
+                if d == dim:
+                    indfl.append(slice(lw-1, None, -1))
+                    indgl.append(slice(0, lw, 1))
+                    indfr.append(slice(-1, -(lw+1), -1))
+                    indgr.append(slice(-lw, None, 1))
+                else:
+                    indfl.append(slice(0, None, 1))
+                    indgl.append(slice(lw, -lw, 1))
+                    indfr.append(slice(0, None, 1))
+                    indgr.append(slice(lw, -lw, 1))
+            if isinstance(f, np.ndarray):
+                g.data[tuple(indgl)] = f[tuple(indfl)]
+                g.data[tuple(indgr)] = f[tuple(indfr)]
+            elif isinstance(f, dv.Function):
+                g.data[tuple(indgl)] = f.data[tuple(indfl)]
+                g.data[tuple(indgr)] = f.data[tuple(indfr)]
+            else:
+                raise NotImplementedError
+        else:
+            raise ValueError("Mode not available")
+
+        return
+
+    def subfloor(f, g):
```

Comment:
Probably don't need this with the right sizes + Operator rather than python fill_data

Reply:
Yeah, probably, will test.